File: newton-4.00/applications/toolsAndWrapers/newtonPy/newtonWorld.py
# Copyright (c) <2003-2021> <Newton Game Dynamics>
 
# This software is provided 'as-is', without any express or implied
# warranty. In no event will the authors be held liable for any damages
# arising from the use of this software.
 
# Permission is granted to anyone to use this software for any purpose,
# including commercial applications, and to alter it and redistribute it
# freely

import bpy
import newton

class NewtonWorldProperties(bpy.types.PropertyGroup):
    solverSubSteps: bpy.props.IntProperty(name= "solver substeps", description="number of solver sub step per ticks", default = 2, min=0, max=8)
    solverIterations: bpy.props.IntProperty(name= "solver iterations", description="Set the number of solver iterations per sub step", default = 4, min=4, max=16)
    
    #my_float_vector : bpy.props.FloatVectorProperty(name= "Scale", soft_min= 0, soft_max= 1000, default= (1,1,1))
    #
    #my_enum: bpy.props.EnumProperty(
    #    name= "Enumerator / Dropdown",
    #    description= "sample text",
    #    items= [('OP1', "Add Cube", ""),
    #            ('OP2', "Add Sphere", ""),
    #            ('OP3', "Add Suzanne", "")
    #    ]
    #)

class NewtonWorld(bpy.types.Object):
    """create an interface to the newton workd"""

    def __init__(self, object):
        #print ("create manager one once")
        self.name = 'newton_world'
        self.world = newton.ndWorld()

        #self.iterations = self.world.GetSolverIterations()
        #self.SetIterations(10)

    def SetIterations (self, iterations):
        self.world.SetSolverIterations(iterations)

class NewtonWorldCreateHomeObject(bpy.types.Operator):
    """Creates a newton world home"""
    bl_label = 'create newton world'
    bl_idname = 'view3d.newton_world_create_home'
    bl_description = "create newton world"

    def execute(self, context):
        scene = context.scene
        selectedObjec = context.active_object
        bpy.ops.mesh.primitive_cube_add(size=1, enter_editmode=False, align='WORLD', location=(0, 0, 0), scale=(1, 1, 1))
        if selectedObjec == context.active_object:
            print ('change to [object mode] operator canceled')
            return {'CANCELLED'}

        context.active_object.name = 'newtonHome'
        world = NewtonWorld(context.active_object)

        print ("test class NewtonWorld")
        # this line works
        world.SetIterations(10)

        # this line fail in blender but not in python
        scene.newton_world = world
        scene.newton_world.SetIterations(10)

        return {'FINISHED'}

class NewtonWorldCreate(bpy.types.Operator):
    """Creates a newton world"""
    bl_label = 'create newton world'
    bl_idname = 'view3d.newton_world_create'
    bl_description = "create a newton world"

    def execute(self, context):
        scene = context.scene

        # this does not works.
        #scene.newton_world = bpy.data.objects.new('newton_world', None) 
        world = NewtonWorld(context.active_object)
        scene.newton_world = world
        return {'FINISHED'}

class NewtonWorldDestroy(bpy.types.Operator):
    """Destroy a newton world"""
    bl_label = 'delete newton world'
    bl_idname = 'view3d.newton_world_destroy'
    bl_description = "destroy a newton world"

    def execute(self, context):
        scene = context.scene

        scene.newton_world.name = 'newtonHome'
        scene.newton_world = None
        return {'FINISHED'}

class NewtonWorldSetProperty(bpy.types.Operator):
    """newton world set engine properties"""
    bl_label = 'newton world set property'
    bl_idname = 'view3d.newton_world_set_property'
    bl_description = "newton world set property"

    def findWorld(self):
        col = bpy.data.collections.get("Collection")
        if col:
            for obj in col.objects:
                if obj.name == 'newton_world':
                    print ("esta es una pura mierda 0")
                    return obj
        return None


    def execute(self, context):
        scene = context.scene
        world = self.findWorld()
        propertyGroup = scene.newton_world_properties
        world.SetIterations(propertyGroup.solverIterations)
        return {'FINISHED'}

File: newton-4.00/applications/toolsAndWrapers/newtonPy/newtonWorld.cpp
/* Copyright (c) <2003-2021> <Newton Game Dynamics>
* 
* This software is provided 'as-is', without any express or implied
* warranty. In no event will the authors be held liable for any damages
* arising from the use of this software.
* 
* Permission is granted to anyone to use this software for any purpose,
* including commercial applications, and to alter it and redistribute it
* freely
*/

#include "newtonStdafx.h"
#include "newtonWorld.h"

class ndContactCallback : public ndContactNotify
{
	public:
	enum ndMaterialUserIDs
	{
		m_default = 0,
		m_dedris = 100,
	};

	enum ndMaterialFlags
	{
		m_playSound = 1 << 0,
		//m_debrisBody = 1 << 1,
	};

	class ndMaterailKey
	{
		public:
		ndMaterailKey()
			:m_key(0)
		{
		}

		ndMaterailKey(dUnsigned64 low, dUnsigned64 high)
			:m_lowKey(dUnsigned32(dMin(low, high)))
			, m_highKey(dUnsigned32(dMax(low, high)))
		{
		}

		bool operator<(const ndMaterailKey& other) const
		{
			return (m_key < other.m_key);
		}

		bool operator>(const ndMaterailKey& other) const
		{
			return (m_key > other.m_key);
		}

		union
		{
			struct
			{
				dUnsigned32 m_lowKey;
				dUnsigned32 m_highKey;
			};
			dUnsigned64 m_key;
		};
	};

	ndContactCallback::ndContactCallback()
		:ndContactNotify()
		//,m_materialMap()
	{
		//m_materialMap.Insert(ndMaterial(), ndMaterailKey(0, 0));
	}


	//virtual ndMaterial& RegisterMaterial(dUnsigned32 id0, dUnsigned32 id1);
	//
	//virtual void OnBodyAdded(ndBodyKinematic* const body) const;
	//virtual void OnBodyRemoved(ndBodyKinematic* const body) const;
	//virtual ndMaterial GetMaterial(const ndContact* const contactJoint, const ndShapeInstance& instance0, const ndShapeInstance& instance1) const;
	//virtual bool OnAabbOverlap(const ndContact* const contactJoint, dFloat32 timestep);
	//virtual void OnContactCallback(dInt32 threadIndex, const ndContact* const contactJoint, dFloat32 timestep);
	//
	//void PlaySoundTest(const ndContact* const contactJoint);
	//
	//dTree<ndMaterial, ndMaterailKey> m_materialMap;
};

NewtonWorld::NewtonWorld()
	:ndWorld()
	,m_nominalTimestep(dFloat32 (1.0f) / NOMINAL_FPS)
{
	ClearCache();
	SetContactNotify(new ndContactCallback);
}

NewtonWorld::~NewtonWorld()
{
}

void NewtonWorld::SetSubSteps(dFloat32 timestep)
{
	timestep = dClamp(timestep, dFloat32(1.0f/120.0f), dFloat32(1.0f / 24.0f)) - 0.001f;
	dInt32 substeps = dInt32 (dFloor(timestep / m_nominalTimestep)) + 1;
	ndWorld::SetSubSteps(substeps);
}

void NewtonWorld::SetTimestep(dFloat32 nominalTimestep)
{
	m_nominalTimestep = dClamp(nominalTimestep, dFloat32(60.0f), dFloat32(600.0f));
}

void NewtonWorld::SetIterations(dInt32 iterations)
{
	iterations = dClamp(iterations, 4, 16);
	SetSolverIterations(iterations);
}

void NewtonWorld::Update(dFloat32 timestep)
{
	timestep = dClamp(timestep, dFloat32(1.0f / 120.0f), dFloat32(1.0f / 24.0f));
	ndWorld::Update(timestep);
}